### ISSUE TITLE
unicode option produces irregular output

### ISSUE THREAD

**Issue by flueterflam** [ho-tex/hyperref#32]
I came across a situation using the `unicode` package option that I believe is a bug.

When loading either `unicode` or `unicode=true` as a package option via `\usepackage`, many of the Doc Info fields are output as strings of hex.  The string of hex is surrounded by < and > (not parentheses, as expected).  For example:

`<</Creator<feff006d0069006b007400650078002c002000780065006c0061007400650078002c0020006d0061006b00650069006e006400650078002c0020006d0061006b00650067006c006f0073007300610072006900650073002c0020006200690062006c006100740065007800200028006200690062006500720029>/Producer(XeTeX 0.99996)/CreationDate(D:20170405214902+10'00')>>`

This seems to apply to most text fields, such as `pdftitle`, `pdfsubject`,`pdfkeywords`, and `pdfcreator`.  It applies whether these fields are loaded as package options also, or via `\hypersetup`.

If `unicode` is set **also** in `\hypersetup`, this behavior still occurs.

However, if you remove `unicode` from the package options and **only** use it in `\hypersetup`, then the text appears normally.  For example:

`<</Creator(unicode bug test)/Producer(XeTeX 0.99996)/CreationDate(D:20170405215804+10'00')>>`

MWE:

    \documentclass{article}

    \usepackage[xetex]{hyperref}%works as expected
    \usepackage[xetex,unicode]{hyperref}%causes issue described above

    \hypersetup{%
	    unicode,%works alone, but with unicode loaded above, it is ignored
	    pdfcreator={unicode bug test},%
    }%

    \begin{document}
    Test
    \end{document}

**Comment by davidcarlisle** [comment]
On 5 April 2017 at 13:03, flueterflam <notifications@github.com> wrote:

> I came across a situation using the unicode package option that I believe
> is a bug.
>
> When loading either unicode or unicode=true as a package option via
> \usepackage, many of the Doc Info fields are output as strings of hex.
> The string of hex is surrounded by < and > (not parentheses, as expected).
> For example:
>
> <</Creator<feff006d0069006b007400650078002c002000780065006c006100740065
> 0078002c0020006d0061006b00650069006e006400650078002c0020006d
> 0061006b00650067006c006f0073007300610072006900650073002c0020
> 006200690062006c006100740065007800200028006200690062006500720029>/Producer(XeTeX
> 0.99996)/CreationDate(D:20170405214902+10'00')>>
>

that is I think the correct thing (the unicode points as hex, starting with
a Byte order mark0 but I am not sure how you are looking at the strings, I
can not see them in this form, the pdf  being compressed, but acrobat at
least shows "unicode bug test" in the document properties, as does pdfinfo
on the command line

Creator:        unicode bug test
Producer:       XeTeX 0.99996
CreationDate:   Wed Apr  5 21:40:32 2017 BST
Tagged:         no
UserProperties: no
Suspects:       no
Form:           none
JavaScript:     no
Pages:          1
Encrypted:      no
Page size:      612 x 792 pts (letter)
Page rot:       0
File size:      2702 bytes
Optimized:      no
PDF version:    1.5

This seems to apply to most text fields, such as pdftitle, pdfsubject,
> pdfkeywords, and pdfcreator. It applies whether these fields are loaded
> as package options also, or via \hypersetup.
>
> If unicode is set *also* in \hypersetup, this behavior still occurs.
>
> However, if you remove unicode from the package options and *only* use it
> in \hypersetup, then the text appears normally. For example:
>
> <</Creator(unicode bug test)/Producer(XeTeX 0.99996)/CreationDate(D:
> 20170405215804+10'00')>>
>

I think that is because the option is essentially ignored in that case as
the package has already committed to using the PDF docstring encoding.


David
[but I'm not that familiar with the details here, Heiko may comment further]


The above pdfinfo output was from the output of

\documentclass{article}

%\usepackage[xetex]{hyperref}%works as expected
\usepackage[unicode]{hyperref}%causes issue described above

\hypersetup{%
    unicode,%works alone, but with unicode loaded above, it is ignored
    pdfcreator={unicode bug test},%
}%

\begin{document}
Test
\end{document}

**Comment by flueterflam** [comment]
Sorry for not clarifying.  Yes, in the PDF viewers that I've used (SumatraPDF and Adobe Acrobat), these fields appear correctly.  So, it is not a problem in that respect.

I am currently compiling my documents in XeLaTeX and am using the `-output-driver="xdvipdfmx -V6 -z0"` option, which is why my documents are uncompressed and I can view it directly.  There is some information about this indicating that xdvipdfmx compresses everything, including either XMP data or Doc Info (I forgot which at the moment), which is the reason I have compression turned off for initial document creation.

I was actually directly viewing the source of the file in a text editor (I was using Notepad++, but checked with several others on my system and see the same result).  The text that I pasted as output was the direct source of the .pdf file.

It may be that this is not actually a bug.  However, I looked through the documentation of `hyperref` and I couldn't find any indication of this behavior.  If it is not a bug, but it *is* required as a package option and *not* as a `\hypersetup` option, then I think it would at least be good to mention it explicitly in the documentation.

**Comment by u-fischer** [comment]
Your first encoding (with the feff006d0069006b007) is the one you get with `pdfencoding=unicode`, the other one is the result of  `pdfencoding=auto`.

The xetex driver contains the switch 

~~~~
\ifx\HyPsd@pdfencoding\HyPsd@pdfencoding@unicode
\else
  \let\HyPsd@pdfencoding\HyPsd@pdfencoding@auto
\fi
~~~~

This means that by default xetex will use `auto` unless you force it into `unicode` by using the package option `unicode` (or `pdfencoding=unicode`).

 
`unicode` in the \hypersetup doesn't "work", it simply doesn't harm as the driver has already been read. 

As you seem to want the auto-encoding: Remove both unicode options and let the xetex driver do what it wants.

**Comment by flueterflam** [comment]
Thank you for the additional details.  When I originally opened this, I hadn't realized that use of the `unicode` option also converted into hex.  However, after David's  comment, I realized this and became more concerned with the documentation indicating the `unicode` option must be used as a package option and is ignored in `\hypersetup`.  Perhaps I missed it somewhere, but I did not notice the documentation stating not to use this option in `\hypersetup`.

I left this issue open hoping for a small update to the documentation so that other users are explicitly clear that `unicode`  should not be used in `\hypersetup`.
